### ISSUE TITLE
Deprecate PropertySet

### ISSUE THREAD

**Issue by samreid** [phetsims/axon#102]
As discussed in #71 and #101, PropertySet creates more problems than it solves and should be deprecated.  The recommended pattern is described here https://github.com/phetsims/axon/issues/71#issuecomment-249094002 

with reset and disposal as described here:
https://github.com/phetsims/axon/issues/71#issuecomment-249294946

If anyone wants to argue that PropertySet should be kept/fixed/etc, that should happen here.  Otherwise, this issue is about how and when we will move away from PropertySet.  Some possibilities:
(a) Replace all PropertySet occurrences now up front, then delete PropertySet immediately.
(b) Leave existing PropertySet occurrences for now, but no new code will use PropertySet
(c) When PhET-iO instrumenting a simulation, refactor away from PropertySet

We also have the possibility to refactor PropertySet as @pixelzoom described in https://github.com/phetsims/phet-io/issues/640 as an intermediate solution.

@jessegreenberg @pixelzoom @aadish @andrewadare @jbphet @jonathanolson please comment below and we'll decide at an upcoming developer meeting.

**Comment by pixelzoom** [comment]
I'm test driving this approach (using Property instead of PropertySet) in models-of-the-hydrogen-atom, where I had 5 PropertySets. See https://github.com/phetsims/models-of-the-hydrogen-atom/commit/534a9c0aa0931bfd53f93b5592e6c50c7a019560 to compare the approaches. Having to implement reset is (as expected) a bit of a pain. But being able to navigate to Properties is wonderful, and this approach looks like it will be much easier to instrument for PhET-iO.

**Comment by samreid** [comment]
@jonathanolson proposed keeping ES5 get/set in https://github.com/phetsims/axon/issues/71#issuecomment-250016697 and I responded in https://github.com/phetsims/axon/issues/71#issuecomment-250059917

**Comment by samreid** [comment]
We have to decide in https://github.com/phetsims/axon/issues/71 before doing anything here.

**Comment by pixelzoom** [comment]
Over in https://github.com/phetsims/axon/issues/71, we decided to move away from PropertySet.  So it's constructor should be annotated as `@deprectated`.

**Comment by pixelzoom** [comment]
PropertySet constructor and its public public functions as now annotated as `@deprecated`. 

Labeling for developer meeting to decide how to proceed with refactoring away from PropertySet.

**Comment by pixelzoom** [comment]
The repositories below use PropertySet.  Shall we make an issue for each repository?

**common code:**
joist
scenery-phet
sun
vegas

**simulations:**
area-builder
arithmetic
balancing-act
balloons-and-static-electricity
bending-light
blackbody-spectrum
build-a-molecule
capacitor-lab-basics
charges-and-fields
circuit-construction-kit-common
color-vision
curve-fitting
energy-skate-park-basics
faradays-law
fluid-pressure-and-flow
forces-and-motion-basics
fraction-comparison
fraction-matcher
friction
gravity-and-orbits
gravity-force-lab-basics
john-travoltage
least-squares-regression
masses-and-springs
molecule-shapes
molecules-and-light
neuron
ohms-law
options-lab
pendulum-lab
projectile-motion
proportion-playground
protein-synthesis
resistance-in-a-wire
rutherford-scattering
seasons
sugar-and-salt-solutions
trig-tour
wave-on-a-string

**Comment by samreid** [comment]
There are 160 occurrences of `PropertySet.call` and 20 occurrences of `new PropertySet`.

**Comment by pixelzoom** [comment]
I removed PropertySet from simula-rasa, blast, chains and vegas.

**Comment by samreid** [comment]
@ariel-phet recommended to create 30+ issues, one per each repo.  Even better: Let's add a note to the "how to instrument" doc to this effect.

@pixelzoom says we should fix the common code that uses propertyset.  @pixelzoom will create issues for that.

@jbphet will tackle ButtonModel.

**Comment by pixelzoom** [comment]
I'm going to reopen this issue until we actually delete PropertySet.

**Comment by jonathanolson** [comment]
Can this be un-tagged for developer meeting?

**Comment by pixelzoom** [comment]
PropertySet has been expunged from common code, and is currently used only in sim-specific repositories.  And we decided that we'll expunge from sim-specific code as we instrument for PhET-iO. So yes, looks like we can remove the developer meeting label.  And I believe that we can also close this issue.

**Comment by samreid** [comment]
Hello,

I added a method to Property.js to help in porting away from PropertySet.  Consider the PropertySet code:

function SomeConstructor(){
  PropertySet.call(this,{name:'test',age:123});
  ...
}

When converting to Property, you can add this check to make sure you catch all ES5 getter setter references:

function SomeConstructor(){
  this.nameProperty = ...
  this.ageProperty = ...
  Property.preventGetSet(this,'name');
  Property.preventGetSet(this,'age');
}

Then any code anywhere else that tries to set variable.name or variable.age will throw an assertion.

Best Regards,
Sam

**Comment by pixelzoom** [comment]
Linked to commits related to `preventGetSet`.   I marked it as `@deprecated`, since it should not appear in production code, and should be deleted once PropertySet is gone.

**Comment by samreid** [comment]
That seems good, should we also recommend putting it behind asserts so it is stripped for production code?

```js
assert && Property.preventGetSet(...);
```

**Comment by pixelzoom** [comment]
> ... should we also recommend putting it behind asserts so it is stripped for production code?

Why would it be in production code?

**Comment by pixelzoom** [comment]
> ... should we also recommend putting it behind asserts so it is stripped for production code?

Why would it be in production code? It's presumably a debugging tool.  Developers should add it, run tests, then remove it.

**Comment by pixelzoom** [comment]
How about moving `preventGetSet` to PropertySet, instead of Property?  Then it will go away when we delete PropertySet.

**Comment by pixelzoom** [comment]
... and developers will be encourage to remove it from their code so that they can remove the `var PropertySet = require(...)` statement.

**Comment by pixelzoom** [comment]
I created "convert from PropertySet to Property" issues for all sims that haven't been published yet. For new sims, I don't see any reason why the conversion shouldn't be done before publishing.

**Comment by pixelzoom** [comment]
Reopening this because @samreid and I seem to still be working on it.

**Comment by pixelzoom** [comment]
Re https://github.com/phetsims/axon/issues/102#issuecomment-265300832...
> How about moving `preventGetSet` to PropertySet, instead of Property? 

Changed my mind - I think it's fine in Property.

**Comment by pixelzoom** [comment]
Before closing this, let's confirm our decisions regarding "conversion of PropertySet to Property" at the next developer meeting. My understanding is:

• Convert new sims before publishing.
• Convert existing sims when they are instrumented for PhET-iO.
• Do not create any new code the uses PropertySet.

**Comment by pixelzoom** [comment]
... and feel free to convert earlier, at developer discretion.

**Comment by mattpen** [comment]
Developers agree with the bullet points above, there was a suggestion to give this to Denzell as a chip-away task.

**Comment by ariel-phet** [comment]
Let's remove what has been done so far, so I can see how many sims are left on this checklist

- [ ] area-builder
- [ ] arithmetic
- [ ] balancing-act
- [ ] balloons-and-static-electricity
- [ ] bending-light
- [ ] blackbody-spectrum
- [ ] build-a-molecule
- [ ] capacitor-lab-basics
- [ ] charges-and-fields
- [ ] circuit-construction-kit-common
- [ ] color-vision
- [ ] curve-fitting
- [ ] energy-skate-park-basics
- [ ] faradays-law
- [ ] fluid-pressure-and-flow
- [ ] forces-and-motion-basics
- [ ] fraction-comparison
- [ ] fraction-matcher
- [ ] friction
- [ ] gravity-and-orbits
- [ ] gravity-force-lab-basics
- [ ] john-travoltage
- [ ] least-squares-regression
- [ ] masses-and-springs
- [ ] molecule-shapes
- [ ] molecules-and-light
- [ ] neuron
- [ ] ohms-law
- [ ] options-lab
- [ ] pendulum-lab
- [ ] projectile-motion
- [ ] proportion-playground
- [ ] protein-synthesis
- [ ] resistance-in-a-wire
- [ ] rutherford-scattering
- [ ] seasons
- [ ] sugar-and-salt-solutions
- [ ] trig-tour
- [ ] wave-on-a-string

**Comment by pixelzoom** [comment]
I updated the above checklist, by removing sims that no longer use PropertySet.

**Comment by jonathanolson** [comment]
- [ ] area-builder
- [ ] arithmetic
- [ ] balancing-act
- [ ] bending-light
- [ ] blackbody-spectrum
- [ ] build-a-molecule
- [ ] capacitor-lab-basics
- [ ] charges-and-fields
- [ ] circuit-construction-kit-common
- [ ] color-vision
- [ ] curve-fitting
- [ ] energy-skate-park-basics
- [ ] faradays-law
- [ ] fluid-pressure-and-flow
- [ ] forces-and-motion-basics
- [ ] fraction-comparison
- [ ] fraction-matcher
- [ ] friction
- [ ] gravity-and-orbits
- [ ] john-travoltage
- [ ] least-squares-regression
- [ ] masses-and-springs
- [ ] molecule-shapes
- [ ] molecules-and-light
- [ ] neuron
- [ ] ohms-law
- [ ] options-lab
- [ ] pendulum-lab
- [ ] projectile-motion
- [ ] proportion-playground
- [ ] protein-synthesis
- [ ] resistance-in-a-wire
- [ ] rutherford-scattering
- [ ] seasons
- [ ] sugar-and-salt-solutions
- [ ] trig-tour

**Comment by pixelzoom** [comment]
I reviewed current usage of PropertySet, and checked off sims that appear to have been converted. Some progress, but still many sims (24) using PropertySet.

**Comment by Denz1994** [comment]
Reference https://github.com/phetsims/projectile-motion/issues/18 for discussion on how most of these can be taken care of.

**Comment by ariel-phet** [comment]
Note to self, delegate

**Comment by samreid** [comment]
Bending Light is checked off after the above commits.

**Comment by samreid** [comment]
CCK basics fixed after the above commit.

**Comment by samreid** [comment]
Protein synthesis complete.

**Comment by samreid** [comment]
Finished sugar-and-salt-solutions above.

**Comment by samreid** [comment]
Energy skate park basics completed, but it was much more complex and hence requires assistance from the QA team when the time is right.  I created an issue https://github.com/phetsims/energy-skate-park-basics/issues/371 for that and I'll check it off the list since it is no longer using PropertySet.

**Comment by samreid** [comment]
The SR sims mentioned in https://github.com/phetsims/axon/issues/102#issuecomment-258203985 are checked off and no longer using PropertySet, so I'll unassign myself.

**Comment by veillette** [comment]
Curve-Fitting was converted from PropertySet to Property back in March. I closed the issue today (https://github.com/phetsims/curve-fitting/issues/104) .

**Comment by ariel-phet** [comment]
@samreid put you back as an assignee since there were a couple of sims in your wheelhouse I had not labelled for you before (apologies)

**Comment by samreid** [comment]
Thanks, it's good to be back!

**Comment by jessegreenberg** [comment]
rutherford-scattering was done today, https://github.com/phetsims/rutherford-scattering/issues/119 is closed.

**Comment by jonathanolson** [comment]
Forgot to mark, handled Pendulum Lab today.

**Comment by samreid** [comment]
I added assignees based on https://github.com/phetsims/axon/issues/102#issuecomment-258203985 but wasn't sure of the GitHub account name for MB

**Comment by jessegreenberg** [comment]
Thanks @samreid, adding @mbarlow12.

**Comment by veillette** [comment]
optics-lab was converted to PropertySet (see https://github.com/phetsims/optics-lab/issues/7)

**Comment by Denz1994** [comment]
Friction and Neuron have been converted to Property from PropertySet.

**Comment by jbphet** [comment]
Area Builder is done.

**Comment by veillette** [comment]
Charges and Fields is done.

**Comment by veillette** [comment]
@jessegreenberg  and I have completed trig-tour

**Comment by pixelzoom** [comment]
I did seasons while running some memory leak tests for another sim. "Some intern" will thank me.

**Comment by mbarlow12** [comment]
Molecules and Light is converted. See https://github.com/phetsims/molecules-and-light/issues/144.

**Comment by jbphet** [comment]
Balancing Act is convert, see https://github.com/phetsims/balancing-act/issues/81.

**Comment by Denz1994** [comment]
Remaining issues in Neuron were addressed. Reference https://github.com/phetsims/neuron/commit/2045f481459b06904ca75fbe861cfc72ed996605

**Comment by samreid** [comment]
@ariel-phet said @samreid should get CCK out for testing before @samreid works on this much more.

**Comment by jbphet** [comment]
Arithmetic is now done.

**Comment by Denz1994** [comment]
Faraday's Law is converted. Reference https://github.com/phetsims/faradays-law/issues/77

**Comment by zepumph** [comment]
@ariel-phet if I end up taking over fliud-pressure-and-flow, I could help in converting it from propertySet if that would be helpful.

**Comment by samreid** [comment]
Fraction matcher is all set.

**Comment by samreid** [comment]
Fluid Pressure and Flow has 75 Properties, so it will take quite a long time to convert. Before starting on it, I'd like @ariel-phet to comment on @zepumph's note:

> @ariel-phet if I end up taking over fliud-pressure-and-flow, I could help in converting it from propertySet if that would be helpful.

@ariel-phet who should work on FPAF?  Or should we split it up?

**Comment by samreid** [comment]
@ariel-phet asked if @zepumph could work on Fluid Pressure and Flow, I'll reassign him in the checkbox comment.

**Comment by samreid** [comment]
@ariel-phet says current priority is "no rush, would be good to be done in a couple of months"

**Comment by pixelzoom** [comment]
Does the preceding comment refer to FPAF, or all remaining sims?

**Comment by pixelzoom** [comment]
8/17/17 dev meeting: Recollection is that @ariel-phet meant "no rush" on all remaining sims, not FPAF specifically.

**Comment by mbarlow12** [comment]
PropertySet removed from [Gravity and Orbits](https://github.com/phetsims/gravity-and-orbits/issues/263).

**Comment by jessegreenberg** [comment]
PropertySet was removed from forces-and-motion-basics.  That was all the sims assigned to @jessegreenberg and @mbarlow12 so removing our assignments.

**Comment by samreid** [comment]
I tackled a few spots in Fluid Pressure and Flow this week.

**Comment by Denz1994** [comment]
PropertySet() was converted for Molecules-Shapes. I don't have any more sims to convert so I am removing my assignment.

**Comment by jonathanolson** [comment]
Removed PropertySet from Build a Molecule. Past self apparently used `new PropertySet( {} )` and `trigger`/`on` for what we now use Emitter for.

Removed my assignment.

**Comment by samreid** [comment]
I finished removing usages of PropertySet then I deleted PropertySet.  I'm going to let Bayes use `preventGetSet` for a few days in Fluid Pressure and Flow before deleting that.  Also leaving this issue open in case there is anything else to discuss about it.  Is the next step to "eliminate" Events? #83 Eliminate is in quotes because we may keep it or a variant of it for scenery node.

**Comment by pixelzoom** [comment]
Agreed, #142 is the reference for converting Events to Emitter.

Looks like our work here is done?

**Comment by zepumph** [comment]
Good work everyone, that was a biggy!
